feat(Stream): switchMap

dilame opened the pull request:
<!--
Before submitting a Pull Request, please ensure you've done the following:

- 📖 Read our Contributing Guide: https://github.com/effect-ts/.github/blob/main/CONTRIBUTING.md
- 📖 Read our Code of Conduct: https://github.com/effect-ts/.github/blob/main/CODE_OF_CONDUCT.md
- 👷‍♀️ Create small PRs. In most cases this will be possible.
- 📝 Use descriptive commit messages.
- ✅ Provide tests for your changes if applicable.
- 📗 If your change requires documentation, please update the relevant documentation.
- 📝 Create a changeset for your changes. This helps in tracking and communicating the changes effectively.
- ⏳ Please be patient! We will do our best to review your pull request as soon as possible.

NOTE: Pull Requests from forked repositories will need to be reviewed by a team member before any CI builds will run.
-->

## Type

<!--
What type of change is this? Please check all applicable.
-->

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

<!--
Please add a brief summary/description of the pull request here.
-->

## Related

<!--
For pull requests that relate or close an issue, please include them below. We like to follow [Github's guidance on linking issues to pull requests](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

For example having the text: "closes #1234" would connect the current pull request to issue 1234 and automatically
close the issue once we merge the pull request.
-->

- Related Issue #
- Closes #

dilame committed:
feat(Stream): switchMap

tim-smart commented:
What is the benefit over `flatMap(f, { switch: true })`?

mikearnaldi commented:
I guess just verbosity, not sure if this is common enough to justify it

dilame commented:
RxJS also do have a [switchMap](https://rxjs.dev/api/index/function/switchMap) operator, and it's widely used.

I decided to make it as a separate operator not just because of verbosity, but because of clarity and visualization.

This is a real code from my project

```ts
export const BinanceWsUnitLayer = Layer.effect(
  BinanceWsStreamTag,
  Effect.gen(function* () {
    const connection = yield* BinanceWsConnectionContextTag;
    const streams = new Map<
      string,
      Stream.Stream<
        unknown,
        | BinanceWsConnectionError
        | BinanceWsResponseError
        | BinanceWsConnectionCloseError
      >
    >();
    return {
      createStream(
        stream: string,
      ): Stream.Stream<
        unknown,
        | BinanceWsConnectionError
        | BinanceWsResponseError
        | BinanceWsConnectionCloseError
      > {
        return Stream.suspend(() => {
          debug(`Requested stream "${stream}"`);
          let stream$ = streams.get(stream);
          if (stream$) {
            return stream$;
          }
          stream$ = connection.pipe(
            Stream.flatMap(
              (ws) =>
                Effect.request(
                  BinanceWsSubscribeRequest({ stream, ws }),
                  RequestResolver.contextFromEffect(
                    BinanceWsSubscribeRequestResolver,
                  ),
                ),
              { switch: true, concurrency: 1 },
            ),
            Stream.flatten(),
          );
          streams.set(stream, stream$);
          return stream$;
        });
      },
    } satisfies BinanceWsStream;
  }),
);
```

It is not really easy to notice the `options` parameter for `flatMap`. When you fluently read this code, you only see `Stream.flatMap`, but default `flatMap` behaviour is different. In addition, there are no clear documentation about what does `switch` option means. On the site it is written

> if the order of concatenation is not important for us, we can use the `switch` option.

And it doesn't describes this concept. 

Actually, i thinks that the `switch` option changes the behaviour dramatically enough to extract it to the separate operator with it's own documentation and example, which i also provided in this PR.

mikearnaldi commented:
@tim-smart what should we do? I am kind of in favour of accepting this given the semantic specificity

tim-smart commented:
> @tim-smart what should we do? I am kind of in favour of accepting this given the semantic specificity

I think we can add it back as `flatMapSwitch` (what it was called before we consolidated the apis).

We would then mark the `switch` option as depreciated and remove it in 4.0?

dilame committed:
feat(Stream): switchMap (since 3.6.0)

dilame commented:
@mikearnaldi done

dilame commented:
/rebase

ecyrbe committed:
Feature: add @effect/sql-kysely package (#3017)

gcanti committed:
sql-kysely: fix error handling in internal executeCommit (#3271)

ecyrbe committed:
fix(kysely): whitelist compile mehod from proxy

dilame commented:
> @dilame can you pls fix the version in `since` to 3.5 and I'll merge it

Hi! Just wanted to remind that i changed the version in `since` that time to `3.5.0` and just now i changed it to `3.6.0` 😃

fubhy commented:
Closing this as stale. The branch is now conflicted and includes unrelated already-merged sql-kysely commits, with no current owner or active review path. If this API is still wanted, please open a fresh focused PR against the current branch.
